plugin/pkg/admission/priority/admission.go

```diff
@@ -0,0 +1,201 @@
+	if p.client == nil {
+		return fmt.Errorf("%s requires a client", pluginName)
+	}
+	if p.lister == nil {
+		return fmt.Errorf("%s requires a lister", pluginName)
+	}
+	return nil
+}
+
+func (p *priorityPlugin) SetInternalKubeClientSet(client internalclientset.Interface) {
+	p.client = client
+}
+
+func (p *priorityPlugin) SetInternalKubeInformerFactory(f informers.SharedInformerFactory) {
+	priorityInformer := f.Scheduling().InternalVersion().PriorityClasses()
+	p.lister = priorityInformer.Lister()
+	p.SetReadyFunc(priorityInformer.Informer().HasSynced)
+}
+
+var (
+	podResource           = api.Resource("pods")
+	priorityClassResource = api.Resource("priorityclasses")
+)
+
+// Admit checks Pods and PriorityClasses and admits or rejects them. It also resolved the priority of pods based on their PriorityClass.
+func (p *priorityPlugin) Admit(a admission.Attributes) error {
+	operation := a.GetOperation()
+	// Ignore all calls to subresources or resources other than pods.
+	// Ignore all operations other than Create and Update.
+	if len(a.GetSubresource()) != 0 || (operation != admission.Create && operation != admission.Update) {
+		return nil
+	}
+
+	switch a.GetResource().GroupResource() {
+	case podResource:
+		return p.admitPod(a)
+
+	case priorityClassResource:
+		return p.admitPriorityClass(a)
+
+	default:
+		return nil
+	}
+}
+
+func (p *priorityPlugin) admitPod(a admission.Attributes) error {
+	operation := a.GetOperation()
+	pod, ok := a.GetObject().(*api.Pod)
+	if !ok {
+		return errors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
+	}
+	if _, isMirrorPod := pod.Annotations[api.MirrorPodAnnotationKey]; isMirrorPod {
+		return nil
+	}
+	// Make sure that the user has not set `priority` at the time of pod creation.
+	if operation == admission.Create && pod.Spec.Priority != nil {
+		return admission.NewForbidden(a, fmt.Errorf("The integer value of priority must not be provided in pod spec. The system populates the value from the given PriorityClass name"))
+	}
+	var priority int32
+	if len(pod.Spec.PriorityClassName) == 0 {
+		dpc, err := p.findDefaultPriorityClass()
+		if err != nil {
+			return fmt.Errorf("Failed to get default priority class: %v", err)
+		}
+		if dpc != nil {
+			priority = dpc.Value
+		} else {
+			priority = scheduling.DefaultPriorityWhenNoDefaultClassExists
+		}
+	} else {
+		// First try to resolve by system priority classes.
+		priority, ok = SystemPriorityClasses[pod.Spec.PriorityClassName]
+		if !ok {
+			// Now that we didn't find any system priority, try resolving by user defined priority classes.
+			pc, err := p.lister.Get(pod.Spec.PriorityClassName)
+			if err != nil {
+				return fmt.Errorf("Failed to get default priority class %s: %v", pod.Spec.PriorityClassName, err)
+			}
+			if pc == nil {
+				return admission.NewForbidden(a, fmt.Errorf("No PriorityClass with name %v was found", pod.Spec.PriorityClassName))
+			}
+			priority = pc.Value
+		}
```

Comment:
Correct. I think we should first provide a way for admin to authorize certain pods to be created with system priorities and then implement the blocking rule. Probably the authorization mechanism will be priority in resource quota, which will automatically address the blocking logic.
I don't think we should reject any pod with system level priority here. Otherwise, even the system component cannot create such pods.